### PR TITLE
Measure network RTT between Thunderscope and Thunderloop

### LINKED ISSUE #3120: Measure network RTT between Thunderscope and Thunderloop
### Description of the task
We should measure the network RTT (Round-trip time) by storying when we send a `Primitive` proto msg from Robot Communication, and storying when we receive a `RobotStatus` message response back from Thunderloop in Robot Communication. The `RobotStatus` message includes the sequence number of the last received Primitive, as such, we can match the `RobotStatus` msg with a specific `Primitive` msg:

https://github.com/UBC-Thunderbots/Software/blob/7b73536a43b5e154d68e655b8454f834fd235218/src/proto/robot_status_msg.proto#L19-L20

This will give us the overall Primitive message to RobotStatus response time, which also includes the duration of Thunderloop. As a result, to calculate the _network RTT_, we can measure the Primitive-RobotStatus response time, and subtract from it the runtime duration of Thunderloop from when it received the Primitive msg, and when it sent out the RobotStatus msg (should be ~`1.0/CONTROL_LOOP_HZ`, though we can also accurately measure that exact value)

This information can be very useful in telling us if we're having latency problems on our network. This is important since changes in latency may have significant effects on our motion planning and how we execute our AI tasks.

### Acceptance criteria

- [ ] Measure the _Network RTT_ between sent Primitive and received RobotStatus in `RobotCommunication`
- [ ] Visualize value in RobotInfo widgets (similar to the packet loss rate labels).

### PR DESCRIPTION
**Pull request by Andrewyx**
### Description

Adds real-time round-trip time widget to the Robot Diagnostics tool. 

### Testing Done
- [x] Field Testing

Resulting RTT Times on SnowBow:

Ping Tests:
image

Script Tests from https://github.com/itsarune/Software/tree/network_latency_test_main.cpp
image

Live Footage in Robot Diagnostics vs Ping:
https://github.com/UBC-Thunderbots/Software/assets/72371419/ac1ab176-3d2b-4dad-9d77-f7e77e8e6a66

### Resolved Issues

Resolves #3120 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/proto/robot_status_msg.proto [hunk 1]
```diff
@@ -19,6 +19,7 @@ message RobotStatus
     Timestamp time_sent                               = 10;
     uint64 last_handled_primitive_set                 = 11;
     PrimitiveExecutorStatus primitive_executor_status = 12;
+    Timestamp omit_thunderloop_processing_time_sent   = 13;
```
Comment (itsarune) [inline src/proto/robot_status_msg.proto]: its weird to name something as "omit"

I think `thunderloop_processing_time` is sufficient

File: src/software/jetson_nano/services/network/network.cpp [hunk 2]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+     *  - make sure it is in the lock
+     *  - log the new PRIMITIVE SET PROTO within a deque if it is valid (see primitive_tracker.send())
+     *  - update the timestamp on each primitive set to the current epoch time
+     */
+
+    if (primitive_set_tuple_rtt.size() >= 50)
+    {
+        LOG(WARNING) << "Too many primitive sets logged for round-trip calculations, halting log process";
+        return;
+    }
```
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.cpp]: when/why would this happen?

File: src/software/jetson_nano/services/network/network.h [hunk 3]
```diff
@@ -79,4 +102,12 @@ class NetworkService
 
     // track last breakbeam state for sending RobotStatus outside of specified rate
     bool last_breakbeam_state_sent = false;
+
+    /**
+     * Deque of key primitive set information as tuples for round-trip time calculations. Tuples specification:
+     * [0] uint64_t: Primitive Sequence Number
+     * [1] double: System time for when primitive set was received by Thunderloop in seconds
+     * [2] double: Epoch time of primitive set sent time from Thunderscope in seconds
```
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.h]: why do we need to store [2]?

File: src/software/logger/logger.h [hunk 4]
```diff
@@ -3,7 +3,7 @@
 #include <g3sinks/LogRotate.h>
 #include <g3sinks/LogRotateWithFilter.h>
 
-#include <experimental/filesystem>
+//#include <experimental/filesystem>
```
Comment (itsarune) [inline src/software/logger/logger.h]: don't commit this

File: src/software/jetson_nano/services/network/network.cpp [hunk 5]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+     *         - store the new omit_thunderloop_processing_time_sent = time_sent + processing duration
+     */
+
+    uint64_t seq_num = robot_status.last_handled_primitive_set();
+    while (seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        if (std::get<0>(primitive_set_tuple_rtt.front()) == seq_num)
+        {
+            double received_epoch_time_seconds = std::get<1>(primitive_set_tuple_rtt.front());
+            double processing_time_seconds = getCurrentEpochTimeInSeconds() - received_epoch_time_seconds;
```
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.cpp]: why not just store `processing_time_seconds` in the proto?

File: src/proto/robot_status_msg.proto (hunk 1 above)
Comment (itsarune) [reply src/proto/robot_status_msg.proto]: I see why you are doing this now -- because in Thunderscope we don't have the timestamp of the last_handled_primitive_set to do latency calculations on.

I feel that adding a new field to this proto that `last_handled_primitive_set_sent_time` helps clean up the code and logic

**Review by itsarune — COMMENTED** [review]
nice work!

btw: I didn't get a chance to test on saturday.

I have one feedback, I think it's sufficient to send the processing_time_seconds back to thunderscope and thunderscope can handle the latency computation. This would mean tracking the send time of the primitive set in the robot status message too.

File: src/software/thunderscope/common/common_widgets.py [hunk 6]
```diff
@@ -65,7 +65,7 @@ class ColorQLabel(QLabel):
     The label starts off with no color and becomes more Red as the value increases up till the max value
     """
 
-    def __init__(self, min_val: float = 0, max_val: float = 100):
+    def __init__(self, min_val: float = 0, max_val: float = 100, text: str = "P%NA"):
```
Comment (nimazareian) [inline src/software/thunderscope/common/common_widgets.py]: nit: We shouldn't have any specific default value here and the caller should define an `initial_text` value. Also, `text` seems to be a bit general, and it should be added to the function docs.

File: src/software/jetson_nano/services/network/network.cpp [hunk 7]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+     * TODO: Send return RobotStatus msg to Thunderscope
+     *      1) Traverse deque containing all received primitive_sets
+     *      2) If the last_handled_primitive_set for current robot_status is stored in the deque:
+     *         - delete all earlier sets
+     *         - stop & calculate the processing counter
+     *         - store the new omit_thunderloop_processing_time_sent = time_sent + processing duration
+     */
+
+    uint64_t seq_num = robot_status.last_handled_primitive_set();
+    while (seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.cpp]: Nit: We should be checking that the deque isn't empty first
```suggestion
    while (!primitive_set_tuple_rtt.empty() && seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
```
This is because `.back()` has underfined behaviour if there are not elements in the deque.
https://en.cppreference.com/w/cpp/container/deque/back

File: src/software/jetson_nano/services/network/network.h [hunk 8]
```diff
@@ -79,4 +102,12 @@ class NetworkService
...
+     */
+    std::deque<std::tuple<uint64_t, double, double>> primitive_set_tuple_rtt;
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.h]: nit: instead of using a `tuple` and defining each value in a comment, it would be more descriptive to define a new `struct` within this class that has three named fields.

File: src/software/jetson_nano/services/network/network.cpp (hunk 5 above)
Comment (nimazareian) [reply src/software/jetson_nano/services/network/network.cpp]: We could, I think the idea was to store one less double in `RobotStatus`.

File: src/software/jetson_nano/services/network/network.cpp [hunk 9]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+
+            robot_status.mutable_omit_thunderloop_processing_time_sent()->set_epoch_timestamp_seconds(
+                    std::get<2>(primitive_set_tuple_rtt.front())
+                    + processing_time_seconds
+            );
+            return;
+        }
+        primitive_set_tuple_rtt.pop_front();
+    }
+    LOG(WARNING) << "No primitives available for round-trip calculations";
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.cpp]: We should probably remove this log as I think it would spam everytime we would trigger the estop (i.e. stop sending primitives).

File: src/software/jetson_nano/services/network/network.cpp [hunk 10]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+}
+
+double NetworkService::getCurrentEpochTimeInSeconds()
+{
+    return std::chrono::duration<double>
+            (std::chrono::steady_clock::now().time_since_epoch()).count();
+}
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.cpp]: Does this get the time in seconds? 

I think the more descriptive way of getting the current time in seconds would be something like:
https://github.com/UBC-Thunderbots/Software/blob/1528173facd4b72d7fd4265e99da5ca9e26b1e3d/src/proto/message_translation/tbots_protobuf.cpp#L348-L353

File: src/software/jetson_nano/services/network/network.cpp [hunk 11]
```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
                      << "%";
     }
 }
+
+void NetworkService::logNewPrimitiveSet(TbotsProto::PrimitiveSet input)
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.cpp]: nit: 
```suggestion
void NetworkService::logNewPrimitiveSet(const TbotsProto::PrimitiveSet& new_primitive_set)
```

File: src/software/jetson_nano/services/network/network.cpp (hunk 7 above)
Comment (nimazareian) [reply src/software/jetson_nano/services/network/network.cpp]: Do we actually even need the check with the `back`? Doesn't `logNewPrimitiveSet` guarantee that the deque is in increasing sequence number and that the `last_handled_primitive_set` is always smaller than the largest sequence number in the deque (i.e. the element in its back)?

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 12]
```diff
@@ -121,7 +121,16 @@ def __init__(
         self.primitive_loss_rate_label = common_widgets.ColorQLabel(
             max_val=MAX_ACCEPTABLE_PACKET_LOSS_PERCENT
         )
+
+        # Primitive round-trip time label and queue
+        self.primitive_rtt_label = common_widgets.ColorQLabel(
+            max_val=MAX_ACCEPTABLE_MILLISECOND_ROUND_TRIP_TIME,
+            text="MS"
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: Lets make the default value 
```suggestion
            text="RTT:NA"
```

File: src/software/thunderscope/constants.py [hunk 13]
```diff
@@ -84,6 +84,11 @@ class EstopMode(IntEnum):
 # the maximum packet / world loss percent indicated by UI
 MAX_ACCEPTABLE_PACKET_LOSS_PERCENT = 30
 
+# maximum acceptable round trip time value in milliseconds
+MAX_ACCEPTABLE_MILLISECOND_ROUND_TRIP_TIME = 200
```
Comment (nimazareian) [inline src/software/thunderscope/constants.py]: lets make this smaller
```suggestion
# maximum acceptable round trip time value in milliseconds
MAX_ACCEPTABLE_MILLISECOND_ROUND_TRIP_TIME = 100
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 14]
```diff
@@ -351,3 +368,11 @@ def __update_ui(self, robot_status: RobotStatus) -> None:
         )
 
         self.battery_progress_bar.setValue(power_status.battery_voltage)
+
+    def __calculate_average_round_trip_time(self, new_time: float) -> int:
+        self.previous_primitive_rtt_values.append(new_time)
+        average_rtt_time_milliseconds = 0
+        for value in self.previous_primitive_rtt_values:
+            average_rtt_time_milliseconds += value
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: nit: you can probably simplify this to
```suggestion
        sum_rtt_time_milliseconds = sum(self.previous_primitive_rtt_values)
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 15]
```diff
@@ -334,13 +343,21 @@ def __update_ui(self, robot_status: RobotStatus) -> None:
...
         self.primitive_loss_rate_label.set_float_val(
             network_status.primitive_packet_loss_percentage
         )
 
+        self.primitive_rtt_label.set_float_val(
+            self.__calculate_average_round_trip_time(
+                ((time.time() - omit_thunderloop_processing_time_sent.epoch_timestamp_seconds)
+                    * MILLISECONDS_PER_SECOND)
+            )
+        )
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: can we split this nested set of called into one or more variables

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 16]
```diff
@@ -351,3 +368,11 @@ def __update_ui(self, robot_status: RobotStatus) -> None:
         )
 
         self.battery_progress_bar.setValue(power_status.battery_voltage)
+
+    def __calculate_average_round_trip_time(self, new_time: float) -> int:
+        self.previous_primitive_rtt_values.append(new_time)
+        average_rtt_time_milliseconds = 0
+        for value in self.previous_primitive_rtt_values:
+            average_rtt_time_milliseconds += value
+        return int(average_rtt_time_milliseconds/len(self.previous_primitive_rtt_values))
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: nit: this doesn't really matter, but a neat little feature in Python is being able to do integer division using `//`. So you can have: `average_rtt_time_milliseconds // len(self.previous_primitive_rtt_values)`

**Review by nimazareian — COMMENTED** [review]
Excited to see this feature in action! 

I just had a couple of nits and suggestions.

File: src/software/thunderscope/common/common_widgets.py (hunk 6 above)
Comment (Andrewyx) [reply src/software/thunderscope/common/common_widgets.py]: So should I change up the existing calls to ColorQLabel with  P%NA instead (Since currently P%NA is hardcoded into the constructor)?

File: src/software/jetson_nano/services/network/network.cpp (hunk 2 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.cpp]: This is purely here to prevent the possibility of a memory leak since thunderloop will be receiving primitive sets at 60Hz, whereas the "cleanup" of the deque occurs at only 30Hz. In theory, an infinitely growing deque is not possible, but in the slight possibility that some error in thunderloop interferes with the creation of a new robot status, which could then break the cleanup process.

File: src/proto/robot_status_msg.proto (hunk 1 above)
Comment (Andrewyx) [reply src/proto/robot_status_msg.proto]: I was thinking of doing the same, but after discussing with Nima, I believe it would be better to avoid adding further data to be sent over the network.

File: src/software/jetson_nano/services/network/network.cpp (hunk 5 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.cpp]: Yup, since even if we stored the processing time within `RobotStatus`, we would still need to store the additional field containing the time sent of the original primitive set (since this is different from the time sent of the `RobotStatus`)

File: src/software/jetson_nano/services/network/network.cpp (hunk 10 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.cpp]: Yes, it does get the time in seconds. I have no idea why `std::chrono` is so convoluted in its type conversions, but this time is a duration for how long the jetson has been powered on. This is because I could not figure out how to have the jetson connect to the internet to sync its onboard time (I am also not sure if this is possible as we would always be connected to the tbots network). In application, since I am using the delta of the times, it makes no difference. 

TLDR: Yes I will reformat it to the suggested form

File: src/software/jetson_nano/services/network/network.h (hunk 3 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.h]: [2] is the time sent of the primitive set as you mentioned previously, I realized that this differs from the time sent within the `RobotStatus` proto as every primitive set will have a different time sent.

File: src/software/jetson_nano/services/network/network.h (hunk 8 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.h]: Good idea, this was exactly what I was looking for

File: src/software/thunderscope/constants.py (hunk 13 above)
Comment (Andrewyx) [reply src/software/thunderscope/constants.py]: I had to temporarily change this for field testing since I was only able to get results of 200ms+ during my testing. I am fairly certain this is a bug.

File: src/software/thunderscope/common/common_widgets.py (hunk 6 above)
Comment (nimazareian) [reply src/software/thunderscope/common/common_widgets.py]: yes, this class should be agnostic of the specific text it's displaying

File: src/software/thunderscope/constants.py (hunk 13 above)
Comment (nimazareian) [reply src/software/thunderscope/constants.py]: @itsarune has a script for measuring latency that we might be able to use to cross reference your values. Alternatively for quickly getting the RTT, you could use `ping` as well.

File: src/software/thunderscope/robot_diagnostics/robot_info.py (hunk 16 above)
Comment (Andrewyx) [reply src/software/thunderscope/robot_diagnostics/robot_info.py]: I gave this a shot and learned something interesting, apparently the floor operator actually produces a float when one of the operands is a float. https://www.learndatasci.com/solutions/python-double-slash-operator-floor-division/

File: src/software/jetson_nano/services/network/network.h [hunk 17]
```diff
@@ -104,10 +104,16 @@ class NetworkService
...
 
     /**
-     * Deque of key primitive set information as tuples for round-trip time calculations. Tuples specification:
-     * [0] uint64_t: Primitive Sequence Number
-     * [1] double: System time for when primitive set was received by Thunderloop in seconds
-     * [2] double: Epoch time of primitive set sent time from Thunderscope in seconds
+     * Struct of key primitive set information as for round-trip time calculations:
+     * sequence_num: Primitive Sequence Number
+     * thunderloop_recieved_time_seconds: System time for when primitive set was received by Thunderloop in seconds
+     * thunderscope_sent_time_seconds: Epoch time of primitive set sent time from Thunderscope in seconds
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.h]: small nit: you can move these comments to be above the fields within the struct

File: src/software/jetson_nano/services/network/network.h [hunk 18]
```diff
@@ -104,10 +104,16 @@ class NetworkService
...
      */
-    std::deque<std::tuple<uint64_t, double, double>> primitive_set_tuple_rtt;
+    struct RoundTripTime {
+        uint64_t primitive_sequence_num;
+        double thunderscope_sent_time_seconds;
+        double thunderloop_recieved_time_seconds;
+    };
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network/network.h]: It's good practice to initialize all fields in a struct/class in C++ since they are not initialized with anything by default and have garbage values.

File: src/software/thunderscope/robot_communication.py [hunk 19]
```diff
@@ -336,9 +338,13 @@ def __enter__(self) -> "self":
...
     #     this means taking the diff of the current time and the omit_thunderloop_processing_time_sent
     #   - Must update the visualized widget component
-    # def __receive_robot_status(self, robot_status: Message) -> None:
-    #     print((time.time() - robot_status.omit_thunderloop_processing_time_sent.epoch_timestamp_seconds) * 1000)
-    #     self.__forward_to_proto_unix_io(RobotStatus, robot_status)
+    #   - Create new proto here, log its time here
+    def __receive_robot_status(self, robot_status: Message) -> None:
+        self.__forward_to_proto_unix_io(
+            RoundTripTime,
+            RoundTripTime(thunderscope_time_received_seconds=time.time())
```
Comment (nimazareian) [inline src/software/thunderscope/robot_communication.py]: Don't forget to update this to be similar to your commented out code:
```suggestion
            RoundTripTime(thunderscope_time_received_seconds=time.time() - robot_status.omit_thunderloop_processing_time_sent.epoch_timestamp_seconds)
```

Comment (nimazareian) [reply src/software/thunderscope/robot_communication.py]: Or I guess you're doing this at the UI. It's just a bit misleading since the protobuf type is called `RoundTripTime`. I think we could update the types name to be related to Robot Statistics or something such that we could also extend in the future with more values.

File: src/software/thunderscope/robot_diagnostics/robot_view.py [hunk 20]
```diff
@@ -133,13 +134,10 @@ def refresh(self) -> None:
         Until the buffer is empty
         """
         robot_status = self.robot_status_buffer.get(block=False, return_cached=False)
+        round_trip_time = self.round_trip_time_buffer.get(block=False, return_cached=False)
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_view.py]: this name should also update to reflect that this is the receive time and not RTT

File: src/software/thunderscope/robot_diagnostics/robot_info.py (hunk 16 above)
Comment (nimazareian) [reply src/software/thunderscope/robot_diagnostics/robot_info.py]: Interesting, good find!

**Review by nimazareian — COMMENTED** [review]
A couple more comments

File: src/proto/robot_status_msg.proto (hunk 1 above)
Comment (Andrewyx) [reply src/proto/robot_status_msg.proto]: rename in progress `adjusted_time_sent`

File: src/software/jetson_nano/services/network/network.cpp (hunk 2 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.cpp]: Updated to scale with Thunderscope refresh rate

File: src/software/thunderscope/robot_diagnostics/robot_view.py [hunk 21]
```diff
@@ -130,9 +134,17 @@ def refresh(self) -> None:
...
+        while robot_status is not None and round_trip_time is not None:
+            self.robot_view_widgets[robot_status.robot_id].update(
+                robot_status, round_trip_time
+            )
             robot_status = self.robot_status_buffer.get(
                 block=False, return_cached=False
             )
+            round_trip_time = self.round_trip_time_buffer.get(
+                block=False, return_cached=False
+            )
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_view.py]: This while loop currently keeps updating the widgets for every robot_status-round_trip_time pair, and only the last pair ends up actually being shown on Thunderscope since all of the previous ones are instantly overwritten. So we could probably improve it by having the while loop go through the entirety of both buffers (probably better to do a separate while loop per buffer too so they both get emptied out even if somehow they are not the same size). Then `if robot_status is not None and round_trip_time is not None` we could `.update` the widgets.

**Review by nimazareian — COMMENTED** [review]
I think everything looks good. Just had a nit and pending your testing tmr.

**Review by nimazareian — DISMISSED** [review]
:racehorse:

File: src/software/jetson_nano/services/network/network.h [hunk 22]
```diff
@@ -79,4 +104,16 @@ class NetworkService
...
     // track last breakbeam state for sending RobotStatus outside of specified rate
     bool last_breakbeam_state_sent = false;
+
+    struct RoundTripTime
+    {
+        // Primitive Sequence Number
+        uint64_t primitive_sequence_num = 0;
+        // System time for when primitive set was received by Thunderloop in seconds
+        double thunderscope_sent_time_seconds = 0;
+        // Epoch time of primitive set sent time from Thunderscope in seconds
```
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.h]: these docs seem mixed up?

File: src/software/thunderscope/robot_communication.py [hunk 23]
```diff
@@ -294,11 +294,11 @@ def __enter__(self) -> "self":
         for RobotStatus, RobotLogs, and RobotCrash msgs, and multicast sender for PrimitiveSet
 
         """
-        # Create the multicast listeners
+
```
Comment (itsarune) [inline src/software/thunderscope/robot_communication.py]: accidentally deleted a comment

Comment (Andrewyx) [reply src/software/thunderscope/robot_communication.py]: fixed

File: src/software/jetson_nano/services/network/network.h (hunk 22 above)
Comment (Andrewyx) [reply src/software/jetson_nano/services/network/network.h]: yup, eagle eyes, fixed it now

File: src/software/thunderscope/robot_diagnostics/robot_view.py [hunk 24]
```diff
@@ -68,14 +68,17 @@ def robot_status_expand(self) -> None:
 
         self.robot_status.toggle_visibility()
 
-    def update(self, robot_status: RobotStatus) -> None:
+    def update(
+        self, robot_status: RobotStatus, round_trip_time: RobotStatistic
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_view.py:72]: looks like `round_trip_time` isn't documented in this function

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 25]
```diff
@@ -351,3 +377,15 @@ def __update_ui(self, robot_status: RobotStatus) -> None:
...
         self.battery_progress_bar.setValue(power_status.battery_voltage)
+
+    def __calculate_average_round_trip_time(self, new_time: float) -> int:
+        """
+        Logs the given round-trip time and calculates the average
+        :param new_time: The new round-trip time to add
+        :return: The mean integer value of the previous round-trip times in milliseconds
+        """
+        self.previous_primitive_rtt_values.append(new_time)
+        average_rtt_time_milliseconds = sum(self.previous_primitive_rtt_values)
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]:
```suggestion
        sum_rtt_time_milliseconds = sum(self.previous_primitive_rtt_values)
```

would probably be a better name

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 26]
```diff
@@ -119,9 +120,23 @@ def __init__(
...
+            label_text="RTT:",
+            initial_value="NA",
+            max_val=MAX_ACCEPTABLE_MILLISECOND_ROUND_TRIP_TIME,
+        )
+        self.previous_primitive_rtt_values = deque(
+            maxlen=MAX_LENGTH_PRIMITIVE_SET_STORE
+        )
+
         self.battery_layout.addWidget(self.primitive_loss_rate_label)
+        self.battery_layout.addWidget(self.primitive_rtt_label)
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: let's rename `battery_layout` to `stats_layout`

File: src/software/jetson_nano/services/network/network.h [hunk 27]
```diff
@@ -51,11 +53,34 @@ class NetworkService
...
+     */
+    double getCurrentEpochTimeInSeconds();
+
     // Constants
     static constexpr float PROTO_LOSS_WARNING_THRESHOLD          = 0.1f;
     static constexpr unsigned int ROBOT_STATUS_BROADCAST_RATE_HZ = 30;
     static constexpr double ROBOT_STATUS_TO_THUNDERLOOP_HZ_RATIO =
         ROBOT_STATUS_BROADCAST_RATE_HZ / (THUNDERLOOP_HZ + 1.0);
+    static constexpr unsigned int PRIMITIVE_DEQUE_MAX_SIZE =
+        static_cast<unsigned int>(1500 / ROBOT_STATUS_BROADCAST_RATE_HZ);
```
Comment (itsarune) [inline src/software/jetson_nano/services/network/network.h:85]: I had to stare at this a bit before realizing what this code was doing

add a comment here for people who'll encounter this later, something along the lines of: `// increases size of deque when robot status messages are sent less frequently`

File: src/software/logger/logger.h (hunk 4 above)
Comment (nimazareian) [reply src/software/logger/logger.h]: Following up on this. Could you revert this change? 
Arun also found that we can fix this by adding this line to the end of the `cc_library` of the logger, in its BUILD file: `linkopts = ["-lstdc++fs"]`
https://github.com/UBC-Thunderbots/Software/blob/d40c8b586b4f8a564f295941c31e12f9270f192e/src/software/logger/BUILD#L16-L31

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 28]
```diff
@@ -115,18 +116,32 @@ def __init__(
...
-        self.battery_layout.addWidget(self.primitive_loss_rate_label)
 
-        self.battery_layout.addWidget(self.battery_progress_bar)
-        self.battery_layout.addWidget(self.battery_label)
+        # Primitive round-trip time label and queue
+        self.primitive_rtt_label = common_widgets.ColorQLabel(
+            label_text="RTT:",
+            initial_value="NA",
+            max_val=MAX_ACCEPTABLE_MILLISECOND_ROUND_TRIP_TIME,
+        )
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: Could you add a `min_val` of 10 as well so the color only starts turning red if it's above 10ms?

**Review by nimazareian — COMMENTED** [review]
Tested the latest changes yesterday and it worked great! Good work. :smiley:

**Review by itsarune — COMMENTED** [review]
LGTM pending those final nits

**Review by nimazareian — APPROVED** [review]
:100:

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 29]
```diff
@@ -351,3 +377,13 @@ def __update_ui(self, robot_status: RobotStatus) -> None:
...
+
+    def __calculate_average_round_trip_time(self, new_time: float) -> int:
+        """
+        Logs the given round-trip time and calculates the average
+        :param new_time: The new round-trip time to add
+        :return: The mean integer value of the previous round-trip times in milliseconds
+        """
+        self.previous_primitive_rtt_values.append(new_time)
+        sum_rtt_time_milliseconds = sum(self.previous_primitive_rtt_values)
+        return int(sum_rtt_time_milliseconds / len(self.previous_primitive_rtt_values))
```
Comment (williamckha) [inline src/software/thunderscope/robot_diagnostics/robot_info.py:389]: fyi you could use `statistics.mean()`

File: src/software/jetson_nano/services/network/network.cpp [hunk 30]
```diff
@@ -74,3 +77,61 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+        // If the proto is older than the last received proto, then ignore it
+        return;
+    }
+
+    NetworkService::RoundTripTime current_round_trip_time;
+    current_round_trip_time.primitive_sequence_num = new_primitive_set.sequence_number();
+    current_round_trip_time.thunderscope_sent_time_seconds =
+        new_primitive_set.time_sent().epoch_timestamp_seconds();
+    current_round_trip_time.thunderloop_recieved_time_seconds =
+        getCurrentEpochTimeInSeconds();
```
Comment (williamckha) [inline src/software/jetson_nano/services/network/network.cpp:102]: Easier to read using brace initialization
```suggestion
    NetworkService::RoundTripTime current_round_trip_time{
        .primitive_sequence_num = new_primitive_set.sequence_number(),
        .thunderscope_sent_time_seconds = new_primitive_set.time_sent().epoch_timestamp_seconds(),
        .thunderloop_recieved_time_seconds = getCurrentEpochTimeInSeconds()};
```

**Review by williamckha — APPROVED** [review]
LGTM! had two comments but they are not important
